src/YOLO/index.js

```diff
@@ -0,0 +1,149 @@
+  constructor(video = null, options = {}) {
+    this.filterBoxesThreshold = options.filterBoxesThreshold || DEFAULTS.filterBoxesThreshold;
+    this.IOUThreshold = options.IOUThreshold || DEFAULTS.IOUThreshold;
+    this.classProbThreshold = options.classProbThreshold || DEFAULTS.classProbThreshold;
+    this.inputDim = options.inputDim || DEFAULTS.inputDim;
+    this.imageSize = 416;
+    this.modelReady = false;
+    this.videoReady = false;
+    this.isPredicting = false;
+
+    this.loadModel();
+
+    if (video instanceof HTMLVideoElement) {
+      this.video = processVideo(video, this.imageSize, () => {
+        this.videoReady = true;
+      });
+    }
+  }
+
+  async loadModel() {
+    this.model = await tf.loadModel(URL);
+    this.modelReady = true;
+  }
+
+  async detect(inputOrCallback, cb = null) {
+    if (this.modelReady && this.videoReady && !this.predicting) {
+      let imgToPredict;
+      let callback;
+      this.isPredicting = true;
+
+      if (inputOrCallback instanceof HTMLImageElement || inputOrCallback instanceof HTMLVideoElement) {
+        imgToPredict = inputOrCallback;
+        callback = cb;
+      } else {
+        imgToPredict = this.video;
+        callback = inputOrCallback;
+      }
+
+      const input = imgToTensor(imgToPredict);
+
+      const [allBoxes, boxConfidence, boxClassProbs] = tf.tidy(() => {
+        const activation = this.model.predict(input);
+        const [boxXY, boxWH, bConfidence, bClassProbs] = head(activation, ANCHORS, 80);
+        const aBoxes = boxesToCorners(boxXY, boxWH);
+        return [aBoxes, bConfidence, bClassProbs];
+      });
+
+      const [boxes, scores, classes] = await filterBoxes(allBoxes, boxConfidence, boxClassProbs, this.filterBoxesThreshold);
+
+      // If all boxes have been filtered out
+      if (boxes == null) {
+        return [];
+      }
+
+      const width = tf.scalar(this.inputDim);
+      const height = tf.scalar(this.inputDim);
+      const imageDims = tf.stack([height, width, height, width]).reshape([1, 4]);
+      const boxesModified = tf.mul(boxes, imageDims);
+
+      const [preKeepBoxesArr, scoresArr] = await Promise.all([
+        boxesModified.data(), scores.data(),
+      ]);
+
+      const [keepIndx, boxesArr, keepScores] = nonMaxSuppression(
+        preKeepBoxesArr,
+        scoresArr,
+        this.IOUThreshold,
+      );
+
+      const classesIndxArr = await classes.gather(tf.tensor1d(keepIndx, 'int32')).data();
+
+      const results = [];
+
+      classesIndxArr.forEach((classIndx, i) => {
+        const classProb = keepScores[i];
+        if (classProb < this.classProbThreshold) {
+          return;
+        }
+
+        const className = CLASS_NAMES[classIndx];
+        let [y, x, h, w] = boxesArr[i];
+
+        y = Math.max(0, y);
+        x = Math.max(0, x);
+        h = Math.min(416, h) - y;
+        w = Math.min(416, w) - x;
+
+        const resultObj = {
+          className,
+          classProb,
+          x: x / 416,
```

Comment:
Should this pick up on the dimensions if they've been changed from the default?

Reply:
Not sure if I understood your question, but this class creates a unique video instance (not visible) with the default size that YOLO needs. So you can have a video of any size and then the results are send as percentages. This means in the sketch you need to do something like this: 

```javascript
 rect(object.x*width, object.y*height, object.w*width, object.h*height);
```

Reply:
Ah, I see. I wasn't sure if `416` in `DEFAULTS.inputDim` from [line 31](https://github.com/ml5js/ml5-library/pull/117/files#diff-a0ffea7193862966598d0692429abff3R31) was something that might change and then would have to be matched later. Perhaps a named constant for the fixed YOLO dimensions would make the code more readable?

This is a minor thing, happy to leave as is if I'm being overly picky!

Reply:
yes, that needs to be constant. will update!